client-mr/src/main/java/org/apache/hadoop/mapred/SortWriteBufferManager.java

```diff
@@ -0,0 +1,346 @@
+    keySerializer.open(buffer);
+    valSerializer.open(buffer);
+    long start = buffer.getDataLength();
+    keySerializer.serialize(key);
+    valSerializer.serialize(value);
+    long end = buffer.getDataLength();
+    long length = end - start;
+    if (length > maxMemSize) {
+      throw new RssException("record is too big");
+    }
+    buffer.addRecord(key, start, end);
+    memoryUsedSize.addAndGet(length);
+    if (memoryUsedSize.get() > maxMemSize * memoryThreshold) {
+      sendBuffersToServers();
+    }
+    mapOutputRecordCounter.increment(1);
+    mapOutputByteCounter.increment(length);
+    return;
+  }
+
+  // Only for test
+  void sendBuffersToServers() {
+    waitSendBuffers.sort(new Comparator<SortWriteBuffer<K, V>>() {
+      @Override
+      public int compare(SortWriteBuffer<K, V> o1, SortWriteBuffer<K, V> o2) {
+        return o1.getDataLength() - o2.getDataLength();
+      }
+    });
+    int sendSize = batch;
+    if (batch > waitSendBuffers.size()) {
+      sendSize = waitSendBuffers.size();
+    }
+    List<SortWriteBuffer<K, V>> selectBuffers = Lists.newArrayList();
+    Iterator<SortWriteBuffer<K, V>> iterator = waitSendBuffers.iterator();
+    while (iterator.hasNext()) {
+      selectBuffers.add(iterator.next());
+      iterator.remove();
+    }
+    List<ShuffleBlockInfo> shuffleBlocks = Lists.newArrayList();
+    for (SortWriteBuffer buffer : selectBuffers) {
+      buffers.remove(buffer.getPartitionId());
+      ShuffleBlockInfo block = createShuffleBlock(buffer);
+      shuffleBlocks.add(block);
+      allBlockIds.add(block.getBlockId());
+      if (!partitionToBlocks.containsKey(block.getPartitionId())) {
+        partitionToBlocks.putIfAbsent(block.getPartitionId(), Lists.newArrayList());
+      }
+      partitionToBlocks.get(block.getPartitionId()).add(block.getBlockId());
+    }
+    sendExecutorService.submit(new Runnable() {
+      @Override
+      public void run() {
+        long size = 0;
+        try {
+          for (ShuffleBlockInfo block : shuffleBlocks) {
+             size += block.getFreeMemory();
+          }
+          inSendListBytes.addAndGet(size);
+          SendShuffleDataResult result = shuffleWriteClient.sendShuffleData(appId, shuffleBlocks);
+          successBlockIds.addAll(result.getSuccessBlockIds());
+          failedBlockIds.addAll(result.getFailedBlockIds());
+        } catch (Throwable t) {
+          LOG.warn("send shuffle data exception ", t);
+        } finally {
+          try {
+            memoryLock.lock();
+            memoryUsedSize.addAndGet(-size);
+            inSendListBytes.addAndGet(-size);
+            full.signalAll();
+          } finally {
+            memoryLock.unlock();
+          }
+        }
+      }
+    });
+  }
+
+  public void waitSendFinished() {
+    while (!waitSendBuffers.isEmpty()) {
+      sendBuffersToServers();
+    }
+    long start = System.currentTimeMillis();
+    long commitDuration = 0;
+    if (!isMemoryShuffleEnabled) {
```

Comment:
I prefer to remove storage types without memory like: local, hdfs, local_hdfs
so, sendCommit() can be avoid

Reply:
OK, it's another pr.